core/state/pruner/pruner.go

```diff
@@ -0,0 +1,517 @@
+	datadir       string
+	trieCachePath string
+	headHeader    *types.Header
+	snaptree      *snapshot.Tree
+}
+
+// NewPruner creates the pruner instance.
+func NewPruner(db ethdb.Database, headHeader *types.Header, datadir, trieCachePath string, bloomSize uint64) (*Pruner, error) {
+	snaptree, err := snapshot.New(db, trie.NewDatabase(db), 256, headHeader.Root, false, false, false)
+	if err != nil {
+		return nil, err // The relevant snapshot(s) might not exist
+	}
+	// Sanitize the bloom filter size if it's too small.
+	if bloomSize < 256 {
+		log.Warn("Sanitizing bloomfilter size", "provided(MB)", bloomSize, "updated(MB)", 256)
+		bloomSize = 256
+	}
+	stateBloom, err := newStateBloomWithSize(bloomSize)
+	if err != nil {
+		return nil, err
+	}
+	return &Pruner{
+		db:            db,
+		stateBloom:    stateBloom,
+		datadir:       datadir,
+		trieCachePath: trieCachePath,
+		headHeader:    headHeader,
+		snaptree:      snaptree,
+	}, nil
+}
+
+func prune(maindb ethdb.Database, stateBloom *stateBloom, start time.Time) error {
+	// Extract all node refs belong to the genesis. We have to keep the
+	// genesis all the time.
+	genesisHashset, err := extractGenesis(maindb)
+	if err != nil {
+		return err
+	}
+	// Delete all stale trie nodes in the disk. With the help of state bloom
+	// the trie nodes(and codes) belong to the active state will be filtered
+	// out. A very small part of stale tries will also be filtered because of
+	// the false-positive rate of bloom filter. But the assumption is held here
+	// that the false-positive is low enough(~0.05%). The probablity of the
+	// dangling node is the state root is super low. So the dangling nodes in
+	// theory will never ever be visited again.
+	var (
+		count  int
+		size   common.StorageSize
+		pstart = time.Now()
+		logged = time.Now()
+		batch  = maindb.NewBatch()
+		iter   = maindb.NewIterator(nil, nil)
+
+		rangestart, rangelimit []byte
+	)
+	for iter.Next() {
+		key := iter.Key()
+
+		// All state entries don't belong to specific state and genesis are deleted here
+		// - trie node
+		// - legacy contract code
+		// - new-scheme contract code
+		isCode, codeKey := rawdb.IsCodeKey(key)
+		if len(key) == common.HashLength || isCode {
+			checkKey := key
+			if isCode {
+				checkKey = codeKey
+			}
+			// Filter out the state belongs to the genesis
+			if _, ok := genesisHashset[common.BytesToHash(checkKey)]; ok {
+				continue
+			}
+			// Filter out the state belongs the pruning target
+			if ok, err := stateBloom.Contain(checkKey); err != nil {
+				return err
+			} else if ok {
+				continue
+			}
+			// Filter out the state belongs to the "blacklist". Usually
+			// the root of the "useless" states are contained here.
+			size += common.StorageSize(len(key) + len(iter.Value()))
+			batch.Delete(key)
+
+			if batch.ValueSize() >= ethdb.IdealBatchSize {
+				batch.Write()
+				batch.Reset()
+			}
+			count += 1
+			if count%1000 == 0 && time.Since(logged) > 8*time.Second {
```

Comment:
I don;t think there's any point in checking `count % 1000 == 0`. If we have the 8 second limit, that should really be fine. An extra condition just makes the log less deterministic (e.g. if the disk is busy).